catalog/controllers/auth.js

```diff
@@ -0,0 +1,63 @@
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken')
+
+const User = require('../models/user');
+const defaultError = require('../util/default-error');
+
+exports.signup = async (req, res, next) => {
+  try {
+    const username = req.body.username
+    const email = req.body.email;
+    const password = req.body.password;
+    const hashedPassword = await bcrypt.hash(password, 12);
+    const user = new User({
+      email,
+      username,
+      password: hashedPassword
+    });
+    const savedUser = await user.save();
+    res.status(201).json({
+      message: 'User successfully created.',
+      user: {
+        userId: savedUser._id,
+        username: savedUser.username
+      }
+    });
+  } catch (err) {
+    next(defaultError(err));
+  }
+};
+
+exports.login = async (req, res, next) => {
+  const email = req.body.email;
+  const password = req.body.password;
+  try {
+    const user = await User.findOne({ email: email })
+    if (!user) {
+      const error = new Error('Wrong email/password.');
+      error.statusCode = 401;
+      throw error;
+    }
+    const passwordIsValid = await bcrypt.compare(password, user.password);
+    if (!passwordIsValid) {
+      const error = new Error('Wrong email/password.');
+      error.statusCode = 401;
+      throw error;
+    }
```

Comment:
> `bcrypt.compare(password, user.password);` 

has to come after User.findOne() and the check that this specific user exists. Otherwise, we would get an error: 

> `Cannot read property 'password' of null`

when performing bcrypt.compare() right after User.findOne().

Reply:
`User.findOne()` and checking that a specific user actually exists has to come before `bcrypt.compare()`
Otherwise, trying to access `user.password` would throw an error because the user object might not exists. Hence the fact that these two operations have to be done one after the other.